languages/ruby/lib/oso/polar/ffi/query.rb

```diff
@@ -74,6 +75,15 @@ def process_messages
             message.process
           end
         end
+
+        # @return [String]
+        # @raise [FFI::Error] if the FFI call returns an error.
+        def source
+          res = Rust.source(self)
+          raise FFI::Error.get if res.null?
```

Comment:
Judging by your CFFI code in polar-c-api/src/lib.rs, it doesn't look like it's possible for this FFI call to return null

Reply:
If there's a panic it would still return null.